File: lib/delta_ex/operations.ex
defmodule DeltaEx.Operations do
  @moduledoc false
  alias DeltaEx.{Config, Native, Telemetry, Util}

  @spec delete(DeltaEx.uri(), String.t(), keyword()) ::
          :ok | {:error, DeltaEx.error_reason()}
  def delete(uri, predicate, opts \\ []) when is_binary(uri) and is_binary(predicate) do
    Telemetry.span(:delete, %{uri: uri, predicate: predicate}, fn ->
      uri
      |> Native.delete_nif(predicate, Util.fetch_storage_options(opts))
      |> normalize_unit_result()
    end)
  end

  @spec update(DeltaEx.uri(), %{String.t() => String.t()}, String.t(), keyword()) ::
          :ok | {:error, DeltaEx.error_reason()}
  def update(uri, updates, predicate \\ "", opts \\ [])
      when is_binary(uri) and is_map(updates) and is_binary(predicate) do
    Telemetry.span(:update, %{uri: uri, predicate: predicate}, fn ->
      uri
      |> Native.update_nif(
        Util.stringify_keys(updates),
        predicate,
        Util.fetch_storage_options(opts)
      )
      |> normalize_unit_result()
    end)
  end

  @spec vacuum(DeltaEx.t(), keyword()) ::
          {:ok, [DeltaEx.uri()]} | {:error, DeltaEx.error_reason()}
  def vacuum(table, opts \\ []) do
    cfg = Config.vacuum()
    retention_hours = Keyword.get(opts, :retention_hours, Keyword.get(cfg, :retention_hours))
    dry_run = Keyword.get(opts, :dry_run, Keyword.get(cfg, :dry_run, true))

    Telemetry.span(:vacuum, %{dry_run: dry_run}, fn ->
      Native.vacuum_nif(table, retention_hours, dry_run)
    end)
  end

  @spec optimize(DeltaEx.t(), keyword()) :: :ok | {:error, DeltaEx.error_reason()}
  def optimize(table, opts \\ []) do
    z_order_columns = Keyword.get(opts, :z_order)

    Telemetry.span(:optimize, %{z_order: z_order_columns}, fn ->
      table
      |> Native.optimize_nif(z_order_columns)
      |> normalize_unit_result()
    end)
  end

  @spec filesystem_check(DeltaEx.t()) :: :ok | {:error, DeltaEx.error_reason()}
  def filesystem_check(table) do
    table
    |> Native.filesystem_check_nif()
    |> normalize_unit_result()
  end

  @spec restore(DeltaEx.t(), keyword()) :: :ok | {:error, DeltaEx.error_reason()}
  def restore(table, opts \\ []) do
    version = Keyword.get(opts, :version)
    datetime = Keyword.get(opts, :datetime)

    case Native.restore_nif(table, version, datetime) do
      {:error, reason} -> {:error, reason}
      _ -> :ok
    end
  end

  @spec convert_to_delta(DeltaEx.uri(), keyword()) ::
          {:ok, DeltaEx.t()} | {:error, DeltaEx.error_reason()}
  def convert_to_delta(uri, opts \\ []) do
    case Native.convert_to_delta_nif(uri, Util.fetch_storage_options(opts)) do
      {:error, _reason} = error -> error
      table -> {:ok, table}
    end
  end

  @spec add_column(DeltaEx.t(), String.t(), String.t() | atom(), keyword()) ::
          :ok | {:error, DeltaEx.error_reason()}
  def add_column(table, column_name, data_type, opts \\ []) do
    nullable = Keyword.get(opts, :nullable, true)

    case Native.add_column_nif(table, column_name, to_string(data_type), nullable) do
      {:error, reason} -> {:error, reason}
      _ -> :ok
    end
  end

  @spec add_constraint(DeltaEx.t(), String.t(), String.t()) ::
          :ok | {:error, DeltaEx.error_reason()}
  def add_constraint(table, name, expression) do
    case Native.add_constraint_nif(table, name, expression) do
      {:error, reason} -> {:error, reason}
      _ -> :ok
    end
  end

  @spec drop_constraint(DeltaEx.t(), String.t()) ::
          :ok | {:error, DeltaEx.error_reason()}
  def drop_constraint(table, name) do
    case Native.drop_constraint_nif(table, name) do
      {:error, reason} -> {:error, reason}
      _ -> :ok
    end
  end

  defp normalize_unit_result({}), do: :ok
  defp normalize_unit_result(result), do: result
end

File: native/delta_ex_native/src/operations.rs
use crate::error;
use crate::expressions::target_session;
use crate::filesystem::table_builder;
use crate::resource::DeltaTableResource;
use crate::runtime::RUNTIME;
use crate::schema::parse_delta_data_type;
use deltalake::kernel::StructField;
use deltalake::operations::optimize::OptimizeType;
use rustler::{Error, ResourceArc};
use std::collections::HashMap;
use std::sync::RwLock;

#[rustler::nif]
pub fn delete_nif(
    uri: String,
    predicate: String,
    storage_options: Option<HashMap<String, String>>,
) -> Result<(), Error> {
    RUNTIME
        .block_on(async {
            let mut table = table_builder(&uri, storage_options)?.build()?;
            table.load().await?;

            let (ctx, df) = target_session(&table)?;
            let schema = df.schema();
            let expr = ctx
                .state()
                .create_logical_expr(&predicate, schema)
                .map_err(error::delta("Predicate"))?;

            table.delete().with_predicate(expr).await?;
            Ok::<(), deltalake::DeltaTableError>(())
        })
        .map_err(error::nif("Delete"))?;

    Ok(())
}

#[rustler::nif]
pub fn update_nif(
    uri: String,
    updates: HashMap<String, String>,
    predicate: String,
    storage_options: Option<HashMap<String, String>>,
) -> Result<(), Error> {
    RUNTIME
        .block_on(async {
            let mut table = table_builder(&uri, storage_options)?.build()?;
            table.load().await?;

            let (ctx, df) = target_session(&table)?;
            let schema = df.schema();

            let mut update_builder = table.update();

            if !predicate.is_empty() {
                let expr = ctx
                    .state()
                    .create_logical_expr(&predicate, schema)
                    .map_err(error::delta("Predicate"))?;
                update_builder = update_builder.with_predicate(expr);
            }

            for (column, expression) in updates {
                let expr = ctx
                    .state()
                    .create_logical_expr(&expression, schema)
                    .map_err(|e| {
                        deltalake::DeltaTableError::Generic(format!(
                            "Update expression for column '{}': {}",
                            column, e
                        ))
                    })?;
                update_builder = update_builder.with_update(column, expr);
            }

            update_builder.await?;
            Ok::<(), deltalake::DeltaTableError>(())
        })
        .map_err(error::nif("Update"))?;

    Ok(())
}

#[rustler::nif]
pub fn vacuum_nif(
    resource: ResourceArc<DeltaTableResource>,
    retention_hours: Option<u64>,
    dry_run: bool,
) -> Result<Vec<String>, Error> {
    let mut table = resource.table.write().unwrap();

    RUNTIME
        .block_on(async {
            let mut vacuum_op = table.clone().vacuum();
            if let Some(h) = retention_hours {
                vacuum_op = vacuum_op.with_retention_period(chrono::Duration::hours(h as i64));
            }
            vacuum_op = vacuum_op.with_dry_run(dry_run);

            let (new_table, result) = vacuum_op.await?;
            *table = new_table;
            Ok::<Vec<String>, deltalake::DeltaTableError>(vec![format!(
                "Deleted {} files",
                result.files_deleted.len()
            )])
        })
        .map_err(error::nif("Vacuum"))
}

#[rustler::nif]
pub fn optimize_nif(
    resource: ResourceArc<DeltaTableResource>,
    z_order_columns: Option<Vec<String>>,
) -> Result<(), Error> {
    let mut table = resource.table.write().unwrap();

    RUNTIME
        .block_on(async {
            let mut optimize_op = table.clone().optimize();
            if let Some(cols) = z_order_columns {
                optimize_op = optimize_op.with_type(OptimizeType::ZOrder(cols));
            }
            let (new_table, _) = optimize_op.await?;
            *table = new_table;
            Ok::<(), deltalake::DeltaTableError>(())
        })
        .map_err(error::nif("Optimize"))
}

#[rustler::nif]
pub fn filesystem_check_nif(resource: ResourceArc<DeltaTableResource>) -> Result<(), Error> {
    let mut table = resource.table.write().unwrap();

    RUNTIME
        .block_on(async {
            let (new_table, _metrics) = table.clone().filesystem_check().await?;
            *table = new_table;
            Ok::<(), deltalake::DeltaTableError>(())
        })
        .map_err(error::nif("FileSystemCheck"))?;

    Ok(())
}

#[rustler::nif]
pub fn restore_nif(
    resource: ResourceArc<DeltaTableResource>,
    version: Option<i64>,
    datetime: Option<String>,
) -> Result<(), Error> {
    let mut table = resource.table.write().unwrap();

    RUNTIME
        .block_on(async {
            let mut builder = table.clone().restore();
            if let Some(v) = version {
                builder = builder.with_version_to_restore(v as u64);
            } else if let Some(dt_str) = datetime {
                let dt = chrono::DateTime::parse_from_rfc3339(&dt_str)
                    .map_err(error::delta("Invalid datetime"))?
                    .with_timezone(&chrono::Utc);
                builder = builder.with_datetime_to_restore(dt);
            } else {
                return Err(deltalake::DeltaTableError::Generic(
                    "Either version or datetime must be provided".to_string(),
                ));
            }

            let (new_table, _metrics) = builder.await?;
            *table = new_table;
            Ok::<(), deltalake::DeltaTableError>(())
        })
        .map_err(error::nif("Restore"))?;

    Ok(())
}

#[rustler::nif]
pub fn convert_to_delta_nif(
    uri: String,
    storage_options: Option<HashMap<String, String>>,
) -> Result<ResourceArc<DeltaTableResource>, Error> {
    let url = crate::filesystem::parse_uri(&uri).map_err(error::nif_msg)?;

    RUNTIME
        .block_on(async {
            let mut convert = deltalake::operations::convert_to_delta::ConvertToDeltaBuilder::new()
                .with_location(url.to_string());
            if let Some(opts) = storage_options {
                if !opts.is_empty() {
                    convert = convert.with_storage_options(opts);
                }
            }
            let table = convert.await?;

            Ok::<ResourceArc<DeltaTableResource>, deltalake::DeltaTableError>(ResourceArc::new(
                DeltaTableResource {
                    table: RwLock::new(table),
                },
            ))
        })
        .map_err(error::nif("Convert"))
}

#[rustler::nif]
pub fn add_column_nif(
    resource: ResourceArc<DeltaTableResource>,
    column_name: String,
    data_type: String,
    nullable: bool,
) -> Result<(), Error> {
    let mut table = resource.table.write().unwrap();
    let dtype = parse_delta_data_type(&data_type).map_err(error::nif_msg)?;
    let field = StructField::new(column_name, dtype, nullable);

    RUNTIME
        .block_on(async {
            let new_table = table.clone().add_columns().with_fields(vec![field]).await?;
            *table = new_table;
            Ok::<(), deltalake::DeltaTableError>(())
        })
        .map_err(error::nif("Add Column"))?;

    Ok(())
}

#[rustler::nif]
pub fn add_constraint_nif(
    resource: ResourceArc<DeltaTableResource>,
    name: String,
    expression: String,
) -> Result<(), Error> {
    let mut table = resource.table.write().unwrap();

    RUNTIME
        .block_on(async {
            let new_table = table
                .clone()
                .add_constraint()
                .with_constraint(name, expression)
                .await?;
            *table = new_table;
            Ok::<(), deltalake::DeltaTableError>(())
        })
        .map_err(error::nif("Add Constraint"))?;

    Ok(())
}

#[rustler::nif]
pub fn drop_constraint_nif(
    resource: ResourceArc<DeltaTableResource>,
    name: String,
) -> Result<(), Error> {
    let mut table = resource.table.write().unwrap();

    RUNTIME
        .block_on(async {
            let new_table = table
                .clone()
                .drop_constraints()
                .with_constraint(name)
                .await?;
            *table = new_table;
            Ok::<(), deltalake::DeltaTableError>(())
        })
        .map_err(error::nif("Drop Constraint"))?;

    Ok(())
}
